### ISSUE TITLE
[Time Display] Timezone incorrect/Shows showing at wrong times

### ISSUE THREAD

**Issue by Sparky1172** [netmix/radio-station#193]
Hi Folks, having some issues with the shows times feature on the plugin. I have set the time zone correctly but doesn't seem to be displayed in the correct times. 

Any ideas on how to fix?

Using MAC 10.14.6 with Safari 

Thank you

**Comment by majick777** [comment]
@Sparky1172 Can you provide a little more information so that we can try to reproduce this? There are a number of places where times are displayed, can you be more specific which ones are displaying incorrectly?

**Comment by majick777** [comment]
Also it might be helpful to check whether your server timezone, WordPress timezone and Radio Station timezone setting differ and report that here.

**Comment by Sparky1172** [comment]
Hi mijick777, 

Sorry it has taken me so long to get back to you.

I am try to display the current presenter/show that is on.

I have attached some screenshots, the first one shows the times allocated for each show, I reversed the times AM times for PM and vice versa as that seemed to work...For a bit 

<img width="1680" alt="Screen Shot 2020-04-26 at 5 22 36 pm" src="https://user-images.githubusercontent.com/64230124/80300794-00dc8780-87e3-11ea-9c56-c4f91396fc95.png">

You can see that Wake Up with Dr Dan is only on between 6PM and 9PM on Saturdays
you can by my screenshot up in the Top right hand corner it's Sunday 5:23pm

<img width="1680" alt="Screen Shot 2020-04-26 at 5 23 00 pm" src="https://user-images.githubusercontent.com/64230124/80300910-d9d28580-87e3-11ea-8483-ec5999a4ba8a.png">

Hope this give you more information, I have check the correct time zone in Settings, they seem to be correct.

**Comment by majick777** [comment]
Okay, so the display in the Current Show widget... 

I did fix a bug with this recently, so first thing you could try would be to install the latest develop version to see if that solves the problem:
https://github.com/netmix/radio-station/tree/develop

Also can I check if you have a page cache? Sometimes that can cause old content to display.

**Comment by majick777** [comment]
@Sparky1172 Any luck?

**Comment by majick777** [comment]
@Sparky1172 going to close this as we haven't heard back from you, if the problem returns we can re-open the issue.

**Comment by tonyzeoli** [comment]
@Sparky1172 Please make sure to take the new 2.3.1 upgrade, which you will not see in the plugins dashboard. You may have to download it. We had named the develop version 2.3.1, so you won't see the update in your Dashboard, but you may miss some of the newer fixes that were implemented i you downloaded the Develop version and worked with that.

**Comment by majick777** [comment]
@Sparky1172 This issue has come up again so I am reopening this ticket and directing related support issues here...

**Comment by majick777** [comment]
For anyone reading this and affected by this issue can you report in summary the following:
- Version of the plugin used (and if it persists with latest development version 2.3.1.4)
- Whether you have AJAX enabled in the plugin settings or on the relevent widgets or not.
- PHP Timezone, WordPress Timezone and Timezone setting on your computer (used by browser.)

**Comment by majick777** [comment]
Support tickets for reference:
https://wordpress.org/support/topic/timezone-issue-1-hour-out/
https://wordpress.org/support/topic/show-currently-on-air-highlighting-1-hour-early/

**Comment by achanner** [comment]
Hi Majick,

This issue to me seems to only be a problem with the 'On Air Now' calculation.

If I show a countdown to the next show, 'Up Next' knows that the show has started (it shows 'This Show Has Started') but 'On Air Now' shows the previous DJ until we get 1 hour beyond and then the display is correct.

I have added two screenshots of the test page I am working on with the clock displayed so that you can see it in action.

**Plug In Bug 1.jpg** shows my presenter Nikki still on air, but she finished at 18:00 but also shows Tonys 'Show has started' in 'Up Next'
**Plug In Bug 2.jpg** shows the 'Up Next' countdown for the next show displaying correctly.

I am using version 2.3.1 of the Plugin
I am not using AJAX Widget Loading as I dont have that version.
Master Schedule displays all the days shows so the time element is not relevant on this for me?
We are currently in Daylight Saving Time

I will be honest to say that I didnt know there were so many timezones on the site!! - I suspect this could be the fix to my problem as PHP and WP timezones dont add up?

Cheers,

Anthony
![Radio Station Plugin Bug 1](https://user-images.githubusercontent.com/67054713/84895778-c08ade80-b09a-11ea-9243-e41af770b3d5.jpg)
![Radio Station Plugin Bug 2](https://user-images.githubusercontent.com/67054713/84895782-c2ed3880-b09a-11ea-97f3-ba91d26cc5d4.jpg)
![PHP Timezone](https://user-images.githubusercontent.com/67054713/84895797-c680bf80-b09a-11ea-87a1-ec8bd5162fce.JPG)
![Client Timezone](https://user-images.githubusercontent.com/67054713/84895805-c8e31980-b09a-11ea-8395-7e3480bd8148.JPG)
![WP Timezone](https://user-images.githubusercontent.com/67054713/84895813-cb457380-b09a-11ea-8380-695d8df30884.JPG)

**Comment by achanner** [comment]
Hi chaps,

Just for additional feedback on this one - I updated the Wordpress timezone (which was out) today and the 'On Air Now' seems to be working perfectly.

**Comment by majick777** [comment]
@achanner Thanks for the info. This confirms what I found, namely that `current_time` uses WordPress timezone. And `date` uses server time (which is usually the same.) But these leave open possibility of a discrepancy with the plugin timezone value.

So I have added a few time and timezone related functions so that these are handled consistently throughout the plugin. Certainly a deeper rabbit hole than expected, but worth it to know we have a more reliable handling of timezones so these kind of issues don't keep coming up again.

These changes have been added to development version 2.3.1.5 for 2.3.2 release.
https://netmix.com/radio-station/docs/FAQ/#how-do-i-install-the-latest-development-version-for-testing

I will leave this ticket open for any further feedback and testing before 2.3.2 is released.
